feat: Agents can reference Query Parameters

mmalmsten opened the pull request:
## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

mmalmsten committed:
Add queryParameterRef to agent properties

mmalmsten committed:
Add documentation

dwmkerr commented on ark/config/crd/bases/ark.mckinsey.com_agents.yaml:
```diff
@@ -117,6 +117,15 @@ spec:
                           - key
                           type: object
                           x-kubernetes-map-type: atomic
+                        queryParameterRef:
```
we'd need to update `./dist` although that's not clear by default so I'll update the generate function

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -475,6 +475,9 @@ func (r *QueryReconciler) finalize(ctx context.Context, query *arkv1alpha1.Query
 
 func (r *QueryReconciler) executeTarget(ctx context.Context, query arkv1alpha1.Query, target arkv1alpha1.QueryTarget, impersonatedClient client.Client, memory genai.MemoryInterface, tokenCollector *genai.TokenUsageCollector) ([]genai.Message, error) {
+	// Store query in context for access in deeper call stacks
+	ctx = context.WithValue(ctx, "queryContextKey", &query)
```
we could extract this into a constant at the top of the controller

dwmkerr commented on ark/internal/genai/agent_parameters.go:
```diff
@@ -87,5 +93,20 @@ func (a *Agent) resolveValueFrom(ctx context.Context, valueFrom *arkv1alpha1.Val
 		return string(value), nil
 	}
 
+	if valueFrom.QueryParameterRef != nil {
```
for the errors around here we could look at how we raise `Events` for more visibilty, a single event if these scenarios fail?

dwmkerr commented on ark/internal/genai/agent_parameters.go:
```diff
@@ -87,5 +93,20 @@ func (a *Agent) resolveValueFrom(ctx context.Context, valueFrom *arkv1alpha1.Val
+			}
+			return "", fmt.Errorf("query parameter %s not found", valueFrom.QueryParameterRef.Name)
+		}
+		return "", fmt.Errorf("no query context available for parameter %s", valueFrom.QueryParameterRef.Name)
```
could we also unit test this as the params can be complex

dwmkerr committed:
feat: add queryParameterRef support for agent parameters

- Add QueryContextKey constant for passing Query resources through context
- Implement event emission for parameter resolution failures (QueryParameterNotFound, QueryParameterResolutionFailed)
- Support nested valueFrom resolution enabling chains like Agent -> Query -> ConfigMap/Secret
- Add comprehensive unit tests covering all resolution scenarios
- Update all dist CRDs with queryParameterRef field
- Add chainsaw tests for nested resolution scenarios
- Refactor resolveValueFrom to reduce complexity by splitting into smaller functions

Addresses PR #140 review feedback

dwmkerr committed:
fix: add newline at end of test file

dwmkerr committed:
fix: add missing ServiceRef validation in parameter valueFrom sources

The validation was only counting 3 of 4 possible valueFrom sources,
causing false validation errors when queryParameterRef was used.
Added ServiceRef to the source counting and name validation logic.

dwmkerr committed:
Merge branch 'main' into feat/add-agent-parameters

dwmkerr committed:
fix: update query-parameter-ref test to use model conditions instead of phase

Models now use status.conditions instead of status.phase field.

dwmkerr committed:
Merge branch 'feat/add-agent-parameters' of github.com:mckinsey/agents-at-scale-ark into feat/add-agent-parameters
